Only first value published when using Effect.merge() and .cancellable()

**Describe the bug**
I've found that using `Effect.merge()` and `.cancellable()` together can in some situations cause an effect to end after its first value is published, rather than waiting for all values to be published. Here's a reduced test case, with a couple of variations that do not fail as well:

**To Reproduce**

```swift
import Combine 
import ComposableArchitecture 
import XCTest 

class MergeCancelTest: XCTestCase { 
  var cancellables: Set<AnyCancellable> = []

  func testNestedMergeCancellation() {
    // The test passes if you change this to Effect.concatenate
    let effect = Effect<Int, Never>.merge(
      (1...2).publisher
        // The test passes if you add .receive(on: testScheduler) and advance the scheduler before XCTAssertEqual
        .eraseToEffect()
        // The test passes if you remove this .cancellable() call
        .cancellable(id: 1)
    )
    // The test passes if you remove this .cancellable() call
    .cancellable(id: 2)
     
    var output: [Int] = []
    effect
      .sink { output.append($0) }
      .store(in: &cancellables)
     
    XCTAssertEqual(output, [1, 2])
  }
} 
```

I've wrapped this up in a test project which shows the failing case plus the multiple variations mentioned in the comments: [MergeCancel.zip](https://github.com/pointfreeco/swift-composable-architecture/files/7809799/MergeCancel.zip)

**Expected behavior**
Both values (1 and 2) should be passed to the `.sink` closure and the test should pass. Instead it fails with this error:

``` 
MergeCancelTests/MergeCancelTests.swift:32: error: -[MergeCancelTests.MergeCancelTest testNestedMergeCancellation] : XCTAssertEqual failed: ("[1]") is not equal to ("[1, 2]")
``` 

As the comments indicate, I know of three ways to make this test pass: 

1. Change `Effect.merge` to `Effect.concatenate`
2. Change the inner publisher to use a non-immediate scheduler
3. Remove either (or both) of the `.cancellable(id:)` calls

(In my current project I've chosen option 2 to get my tests passing, but it makes the tests more complicated than necessary.)

I'm pretty sure the bug is related to `Effect.cancellable` using a `PassthroughSubject` internally, and subscribing that subject to the publisher before the subject has been connected to any downstream subscribers. But I don't understand it fully, because clearly that works just fine in most cases.

**Environment**
  - swift-composable-architecture version 0.32.0
  - Xcode 13.0
  - Swift 5.5
  - OS: macOS 12.0

--- a/Examples/VoiceMemos/VoiceMemosTests/VoiceMemosTests.swift
+++ b/Examples/VoiceMemos/VoiceMemosTests/VoiceMemosTests.swift
@@ -113,7 +113,7 @@ class VoiceMemosTests: XCTestCase {
     environment.audioRecorder.startRecording = { _ in
       audioRecorderSubject.eraseToEffect()
     }
-    environment.mainRunLoop = .immediate
+    environment.mainRunLoop = self.mainRunLoop.eraseToAnyScheduler()
     environment.temporaryDirectory = { .init(fileURLWithPath: "/tmp") }
     environment.uuid = { UUID(uuidString: "DEADBEEF-DEAD-BEEF-DEAD-BEEFDEADBEEF")! }
 
@@ -124,6 +124,7 @@ class VoiceMemosTests: XCTestCase {
     )
 
     store.send(.recordButtonTapped)
+    self.mainRunLoop.advance(by: 0.5)
     store.receive(.recordPermissionResponse(true)) {
       $0.audioRecorderPermission = .allowed
       $0.currentRecording = .init(
@@ -133,6 +134,7 @@ class VoiceMemosTests: XCTestCase {
       )
     }
     audioRecorderSubject.send(completion: .failure(.couldntActivateAudioSession))
+    self.mainRunLoop.advance(by: 0.5)
     store.receive(.audioRecorder(.failure(.couldntActivateAudioSession))) {
       $0.alert = .init(title: .init("Voice memo recording failed."))
       $0.currentRecording = nil
@@ -189,7 +191,7 @@ class VoiceMemosTests: XCTestCase {
   func testPlayMemoFailure() {
     var environment = VoiceMemosEnvironment.failing
     environment.audioPlayer.play = { _ in Effect(error: .decodeErrorDidOccur) }
-    environment.mainRunLoop = .immediate
+    environment.mainRunLoop = self.mainRunLoop.eraseToAnyScheduler()
 
     let url = URL(string: "https://www.pointfree.co/functions")!
     let store = TestStore(
@@ -278,7 +280,7 @@ class VoiceMemosTests: XCTestCase {
     var environment = VoiceMemosEnvironment.failing
     environment.audioPlayer.play = { _ in .none }
     environment.audioPlayer.stop = { .none }
-    environment.mainRunLoop = .immediate
+    environment.mainRunLoop = self.mainRunLoop.eraseToAnyScheduler()
 
     let store = TestStore(
       initialState: VoiceMemosState(

--- a/Sources/ComposableArchitecture/Effects/Cancellation.swift
+++ b/Sources/ComposableArchitecture/Effects/Cancellation.swift
@@ -29,7 +29,7 @@ extension Effect {
   ///     canceled before starting this new one.
   /// - Returns: A new effect that is capable of being canceled by an identifier.
   public func cancellable(id: AnyHashable, cancelInFlight: Bool = false) -> Effect {
-    let effect = Deferred { () -> Publishers.PrefixUntilOutput<Publishers.HandleEvents<Self>, PassthroughSubject<Void, Never>> in
+    let effect = Deferred { () -> Publishers.HandleEvents<Publishers.PrefixUntilOutput<Self, PassthroughSubject<Void, Never>>> in
       cancellablesLock.lock()
       defer { cancellablesLock.unlock() }
 
@@ -50,10 +50,11 @@ extension Effect {
         cancellationCancellable
       )
 
-      return self.handleEvents(
-        receiveCompletion: { _ in cancellationCancellable.cancel() },
-        receiveCancel: cancellationCancellable.cancel
-      ).prefix(untilOutputFrom: cancellationSubject)
+      return self.prefix(untilOutputFrom: cancellationSubject)
+        .handleEvents(
+          receiveCompletion: { _ in cancellationCancellable.cancel() },
+          receiveCancel: cancellationCancellable.cancel
+        )
     }
     .eraseToEffect()
 

--- a/Sources/ComposableArchitecture/Effects/Cancellation.swift
+++ b/Sources/ComposableArchitecture/Effects/Cancellation.swift
@@ -29,18 +29,16 @@ extension Effect {
   ///     canceled before starting this new one.
   /// - Returns: A new effect that is capable of being canceled by an identifier.
   public func cancellable(id: AnyHashable, cancelInFlight: Bool = false) -> Effect {
-    let effect = Deferred { () -> Publishers.HandleEvents<PassthroughSubject<Output, Failure>> in
+    let effect = Deferred { () -> Publishers.PrefixUntilOutput<Publishers.HandleEvents<Self>, PassthroughSubject<Void, Never>> in
       cancellablesLock.lock()
       defer { cancellablesLock.unlock() }
 
-      let subject = PassthroughSubject<Output, Failure>()
-      let cancellable = self.subscribe(subject)
+      let cancellationSubject = PassthroughSubject<Void, Never>()
 
       var cancellationCancellable: AnyCancellable!
       cancellationCancellable = AnyCancellable {
         cancellablesLock.sync {
-          subject.send(completion: .finished)
-          cancellable.cancel()
+          cancellationSubject.send(())
           cancellationCancellables[id]?.remove(cancellationCancellable)
           if cancellationCancellables[id]?.isEmpty == .some(true) {
             cancellationCancellables[id] = nil
@@ -52,10 +50,10 @@ extension Effect {
         cancellationCancellable
       )
 
-      return subject.handleEvents(
+      return self.handleEvents(
         receiveCompletion: { _ in cancellationCancellable.cancel() },
         receiveCancel: cancellationCancellable.cancel
-      )
+      ).prefix(untilOutputFrom: cancellationSubject)
     }
     .eraseToEffect()
 

--- a/Tests/ComposableArchitectureTests/EffectCancellationTests.swift
+++ b/Tests/ComposableArchitectureTests/EffectCancellationTests.swift
@@ -286,4 +286,20 @@ final class EffectCancellationTests: XCTestCase {
     scheduler.advance(by: 1)
     XCTAssertNoDifference(expectedOutput, [])
   }
+  
+  func testNestedMergeCancellation() {
+    let effect = Effect<Int, Never>.merge(
+      (1...2).publisher
+        .eraseToEffect()
+        .cancellable(id: 1)
+    )
+    .cancellable(id: 2)
+
+    var output: [Int] = []
+    effect
+      .sink { output.append($0) }
+      .store(in: &cancellables)
+
+    XCTAssertEqual(output, [1, 2])
+  }
 }

